pkg/auth/dependency/interaction/flows/webapp.go

```diff
@@ -63,11 +98,71 @@ func (f *WebAppFlow) AuthenticatePassword(token string, password string) (*WebAp
 		return nil, i.Error
 	}
 
+	return f.afterPrimaryAuthentication(i)
+}
+
+func (f *WebAppFlow) SetupPassword(token string, password string) (*WebAppResult, error) {
+	i, err := f.Interactions.GetInteraction(token)
+	if err != nil {
+		return nil, err
+	}
+
+	err = f.Interactions.PerformAction(i, interaction.StepSetupPrimaryAuthenticator, &interaction.ActionSetupAuthenticator{
+		Authenticator: interaction.AuthenticatorSpec{Type: interaction.AuthenticatorTypePassword},
+		Secret:        password,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = f.Interactions.SaveInteraction(i)
+	if err != nil {
+		return nil, err
+	}
+
+	if i.Error != nil {
+		return nil, i.Error
+	}
+
 	s, err := f.Interactions.GetInteractionState(i)
+	if err != nil {
+		return nil, err
+	} else if s.CurrentStep().Step != interaction.StepCommit {
+		panic("interaction_flow_webapp: unexpected interaction state")
+	}
+
+	attrs, err := f.Interactions.Commit(i)
 	if err != nil {
 		return nil, err
 	}
 
+	switch i.Intent.Type() {
+	case interaction.IntentTypeSignup:
+		// New interaction for logging in after signup
+		i, err = f.Interactions.NewInteractionLogin(&interaction.IntentLogin{
+			Identity: i.Identity.ToSpec(),
+			AuthenticatedAs: &interaction.IntentLoginAuthenticatedAs{
```

Comment:
I think having `AuthenticatedAs` and do the login again after signup is a good move! But seems the reason of session in `CreateSession` will become always `login`? We use the reason for hook, right?

Reply:
Now the original intent type is stored to determine session create reason.